### ISSUE TITLE
Transformer: Attention Is All You Need

### ISSUE THREAD

**Issue by ANGHOOO** [ANGHOOO/PaperReview#2]
Paper : https://arxiv.org/abs/1706.03762
Github : https://github.com/huggingface/transformers/tree/main

오늘 리뷰할 논문은 현재까지도 SOTA 모델 구조의 근간으로 사용이 되는 transformer 모델에 관한 paper인 Attention Is All You Need입니다. 기존의 RNN, CNN 모델 구조를 사용하지 않고 Attention Mechanism을 통해 여러 NLP task에서 SOTA 성능을 달성한 매우 혁신적인 모델입니다. 현재 SOTA 모델 구조의 근간이 되는 모델인만큼 해당 모델의 구조와 동작 방식에 대해 자세히 알고있는것이 추후 여러 논문을 읽는데 필수적이기 때문에 해당 논문은 반드시 읽어야 하는 논문들 중 하나라고 생각됩니다!

## Contribution

Transformer의 주요 contribution은 다음과 같습니다.

- RNN, CNN 모델 구조를 사용하지 않고 Attention Mechanism을 통한 모델 구조
- 계산이 순차적으로 이루어지는 RNN과 달리 병렬적인 계산이 가능하게 하여 학습 효율 증가

## Transformer Architecture

입력값에 대해 전체 context의 정보를 담고있는 값으로 바꿔주는 Encoder와 번역할 문장을 생성해주는 역할을 하는 Decoder 구조로 나뉘어 집니다. Query, Key, Value에 해당하는 행렬을 통해 Attention value를 계산하게 됩니다. 또한 Multi-Head Attention을 통해 여러가지 관점에서 입력 단어에 대해 어떤 부분들을 중요하게 봐야하는지 알 수 있게 됩니다. Decoder 구조에서는 Masked Multi-Head Attention과 Encoder Decoder Multi-Head Attention을 통해 문장을 번역할 때 입력된 문장의 전체 context 정보를 가지고 출력값을 생성하게 하는 것이 가능하게 하였습니다. 따라서 본 논문을 살펴볼때도 Encoder 구조와 Decoder 구조로 나눠서 자세하게 살펴보도록 하겠습니다.
<img width="319" alt="image" src="https://github.com/ANGHOOO/PaperReview/assets/103275370/e9d66e58-c599-4226-aec2-725794f361b1">

## Encoder

<img src = "https://github.com/ANGHOOO/PaperReview/assets/103275370/ef38f885-2d37-427a-94c3-9b206258a19c" width="319">

### Positional Encoding

---

time step별로 계산이 순차적으로 수행되는 RNN 계열 모델과 달리 transformer 구조는 입력이 한 번에 들어와 병렬적으로 계산이 수행되기 때문에 단어의 순서정보를 잃어버리게 된다. 입력 정보의 순서가 중요한 시계열이나 텍스트 데이터에서 입력 정보를 추가해주기 위해 본 논문에서는 Positional Encoding이라는 방법을 제안한다. 
<img src = "https://github.com/ANGHOOO/PaperReview/assets/103275370/7ae2bb20-88eb-43ff-b52b-a2460e15527c">

순서 정보를 추가해주기 위해서 주기함수인 sine 함수와 cosine 함수를 사용하였다.
여기서 pos는 입력 문장에서 embedding 벡터들의 위치를 나타내며, i는 embedding 벡터 내의 차원의 인덱스를 나타낸다.
embedding 벡터 내의 차원 I가 짝수이면 sine 함수, 홀수이면 cosine 함수를 사용해서 순서 정보를 추가한다.
[-1, 1] 사이의 범위를 가지는 sine, cosine 함수를 통해 위치 정보의 값이 더해질 때, embedding 벡터의 값을 크게 변경시키지 않고 위치 정보를 추가해 줄 수 있게 된다. 또한 pos 값에 따라 sine, cosine 함수의 출력값이 달라지기 때문에 동일한 단어의 embedding 벡터라고 할지라도 pos값이 달라지면 다른 값을 가지는 벡터가 되기 때문에 순서 정보를 반영할 수 있게 된다.

### Scaled Dot-Product Attention

---

<img src = "https://github.com/ANGHOOO/PaperReview/assets/103275370/d7603088-d773-4d8d-a007-6b664803e9ac">

여기서 Q, K, V는 각각 Query, Key, Value에 해당한다.
Transformer 구조는 기계번역 task를 수행하기 위해 고안된 모델 구조이기 때문에 특정 단어가 들어왔을 때, 문장에서 어떤 부분을 중점적으로 봐야하는지 계산하기 위해서 위의 mechanism과 같이 Query, Key, Value 행렬을 만들어주는 과정을 거치게 된다.
본 논문에서는 embedding 벡터의 차원을 512로 설정하였고, Query, Key, Value의 차원은 각각 64로 설정하였다.
이떄 64는 embedding 벡터의 차원을 head의 개수로 나눠준 값이 되며 본 논문에서는 head의 개수를 8로 설정하였기 때문에 512 / 8 = 64라는 값이 된다. Query, Key, Value 행렬은 학습 가능한 파라미터로 구성된 Wq, Wk, Wv 행렬과의 곱들로 생성된다.
이 때, Wq, Wk, Wv 행렬들의 가중치 값들은 각각 다르게 초기화 되기 때문에 생성된 Q, K, V 행렬들도 서로 다른 값을 갖게된다.
여기서 Query 행렬에 대한 Attention score를 계산하기 위해서는 입력 문장의 모든 단어에 대한 Key 행렬과의 내적을 통해 Attention score를 구하게 된다. 그 뒤 softmax 연산을 통하여 Attention score의 분포를 구한 뒤 Value 행렬에 가중합 하는 방식으로 최종 Attention value 행렬을 구한다.

### Multi-Head Attention

---

Multi-Head Attention은 각기 다른 가중치를 갖는 Wq, Wk, Wv 행렬을 여러 개 생성하여 여러개의 Attention Value 행렬을 생성하는 것이다. 신화 속의 동물인 케르베로스나 히드라 등을 생각해본다면 여러 개의 머리를 가지고 있기 때문에 각기 다른 시야 정보를 공유하여 공격을 어렵게 한다. 이와 마찬가지로 Multi-Head Attention도 특정 단어를 볼 때 여러 개의 Attention value 행렬을 통해 정보를 공유할 수 있게 되어 문맥의 정보를 더 잘 파악할 수 있게 된다. 마지막에는 이렇게 생성된 Head들을 단순히 concat하고 선형변환을 통해 크기를 맞춰주는 방식으로 진행된다. Multi-Head Attention을 통해 우리의 모델이 특정 단어를 볼 때 문장에서 어떤 단어들에 집중하고 있는지 확인할 수 있게 되었다.
<img src = "https://github.com/ANGHOOO/PaperReview/assets/103275370/b122524d-3141-4084-8c1a-8bf874b71fb5">

### Add & Normalization

---

Encoder block 내부에 Add & Norm 파트가 들어있는 것을 볼 수 있다.
여기서 Add는 ResNet 논문에서 제안한 Identity mapping을 적용한 것이다.
쌓여진 레이어가 그 다음 레이어에 바로 Optimize 되는 것이 아니라 Residual Mapping에 Optimize 되도록 만든다.
또한 추가적인 파라미터도 필요하지 않으며, Shortcut Connection만 추가되기 때문에 복잡한 곱셈 연산도 필요하지 않다는 장점을 지니고 있다. 이를 통해 모델의 레이어를 깊게 쌓아도 학습이 잘 될 수 있도록 하여 현재의 딥러닝 모델 대부분에 해당 방식이 적용되고 있다.

본 논문에서 Norm 파트는 Layer Normalization 방법을 사용한다.
Layer Normalization을 수행하게 개별 데이터 샘플에 대하여 Normalization을 적용하기 때문에 Batch Normalization과 다르게 배치 사이즈에 무관하게 적용이 가능하다.

### Feed Forward

---

Encoder block에서의 연산 과정을 정리해보면 모델의 입력에 대해서 Positional Encoding을 통해 순서 정보를 추가해주는 과정을 거친다.
그 뒤 Encoder block의 sub layer 중 Multi-Head Attention을 구하는 과정을 거치고 이를 통해 Attention score의 정보가 포함된 출력값이 생성된다. 해당 출력값은 sub layer 중 Feed Forward 연산을 거치게 되고 이는 우리가 아는 일반적인 Feed Forward(MLP) 연산과 동일하다. 본 논문에서는 Feed Forward에서 두 개의 레이어를 사용하였으며 가중치 행렬 W1 = (d_model, d_ff)의 크기를 가지며, W2 = (d_ff, d_model)의 크기를 가지게 됩니다. 여기서 d_ff는 2048로 본 논문에서 설정하였다.

## Decoder

<img src = "https://github.com/ANGHOOO/PaperReview/assets/103275370/dc8f4a42-ebea-47c8-bd8c-aa7cddcbbbf9" width = "319">

### Masked Multi-Head Attention

---

Decoder의 역할은 기계번역 task라고 가정하면 번역이 된 문장들을 출력해주는 단계라고 볼 수 있다.
그렇기에 훈련 과정에서의 Decoder의 입력으로는 실제 번역된 문장들이 Encoder와 마찬가지로 병렬적으로 입력으로 들어온다.
실제로 번역이 수행되는 과정을 생각해보면 번역의 대상이 될 문장들이 들어오고 Decoder를 통해 번역된 단어 하나씩을 출력하고 출력된 단어들이 다시 Decoder의 입력으로 들어가서 문장의 마지막임을 나타내는 토큰이 등장할 때까지 Decoder가 출력을 내뱉게 된다.
하지만 위에서 언급했듯이 학습 과정에서 Decoder의 입력으로는 번역된 문장 전체가 병렬적으로 들어가기 때문에 모델의 입장에서는 이후에 번역이 될 문장까지 전부를 보고 학습하는 방식으로 학습이 될 수 있으며 이는 치팅에 해당한다.
따라서 Decoder에서 Multi-Head Attention을 수행할 때는 입력으로 들어온 문장에서 해당 단어 이후의 단어들에 대해서는 Masking을 적용해서 이후 시점의 단어들에 대해서는 학습에 사용되지 못하는 방식을 적용한다. 
Attention score를 계산하는 과정에서 현재 계산하려는 단어 이후에 등장하는 단어들에 대해서는 softmax 연산을 수행하기 전에 -inf 와 같은 아주 작은 값으로 마스킹을 적용해서 softmax 연산 이후의 값이 0이 되도록 만들어 학습에 영향을 미치지 않도록 한다.

### Encoder Decoder Multi-Head Attention

---

기존처럼 동일한 Query, Key, Value 값들을 받는 것이 아닌 Key, Value 값들은 Encoder의 최종 출력으로부터, Query값은 Decoder의 Masked Multi-Head Attention의 출력 결과 값을 사용한다. 이것의 의미를 잘 생각해보면 실제 우리가 번역할 단어(Query)를 출력할 때 번역의 대상이 되는 문장에서 어떤 단어(Key, Value)들을 주의깊게 봐야 할지 보는 방식에 해당한다고 볼 수 있다.
결과적으로 Encoder Decoder Multi-Head Attention에서 Attention score를 계산하게 되면 번역의 대상이 되는 단어들과 번역된 단어들 사이의 Attention score가 구해지게 된다.

### Feed Forward & Prediction

---

이후의 과정은 Encoder와 마찬가지로 Feed Forward 과정을 거치고 최종 출력으로는 학습에 사용된 vocab size만큼을 갖도록 선형변환을 수행한 뒤 softmax 연산을 거쳐 해당 vocab에서 가장 큰 확률값을 가지는 인덱스를 통해 해당 인덱스의 단어를 출력한다.

## Experiment

<img src = "https://github.com/ANGHOOO/PaperReview/assets/103275370/ee046307-2997-446c-84e6-b4b2f2c52073">

<img src = "https://github.com/ANGHOOO/PaperReview/assets/103275370/2253db78-d80a-43b4-bfca-bd5b0dbb44fb">

GPU Memory만 충분하다면 입력의 크기만큼 병렬적으로 연산을 수행하여 학습 효율을 증대시킬 수 있다.
또한 이전 SOTA 모델들과 비교했을 때 BLEU Score가 Transformer (big model)에서 SOTA 성능을 달성하였다.

# The Pros and Cons of Attention

- Long term dependency : 전통적인 RNN 모델들과 LSTM, GRU와 같은 그 변형들은 time step이 길어질수록 앞에 있는 단어의 정보가 뒤에서 나오는 단어들까지 잘 전달되지 못하는 long term dependency 문제가 있었다. Attention Mechanism은 입력 단어의 어떤 step에서든 직접적인 접근이 가능하게 하여 전반적인 문맥의 정보를 보존할 수 있게 되었다.

- Parallelization : 순차적으로 계산을 수행하는 RNN 모델과 달리, transformer와 같은 attention-based 모델에서는 입력 시퀀스의 토큰들을 병렬적으로 한 번에 처리가 가능하게 하였다. 이것은 계산 효율성에서 큰 이점이 있다.

- Interpretability : Attention은 우리의 모델이 출력을 생성할 때 입력의 어떤 부분을 중요하게 보고 있는지 해석이 가능하게 하였다. "Attention map"을 통해 우리의 모델이 "어떻게 생각하는지"에 대해 우리는 이해할 수 있게 되었다.

- Global Context : CNN 계열의 receptive field는 전형적으로 지역적인 정보와 kernel size에 의존하여 잠재적으로 넓은 정보에 대해서는 손실이 일어나게 된다. 그러나 attention을 통해 출력을 생성할 때 각 토큰은 모든 시퀀스의 정보를 받기 때문에 전체적인 문맥의 정보를 보존할 수 있게 된다.

- Improved Performance : Attention-based 모델은 특히 여러 NLP task에서 RNN, CNN 계열의 모델들에 비해 SOTA 성능을 달성하였다.

**논문 리뷰 full version**
https://anghoo-ai.tistory.com/37
